File: Shared/PPPlugInInfo.m
//
//  PPPlugInInfo.m
//  PPMacho
//
//  Created by C.W. Betts on 12/31/12.
//
//

#import "PPPlugInInfo.h"

@interface PPPlugInInfo ()
@property (readwrite, copy) NSString	*plugName;
@property (readwrite, copy) NSString	*authorName;
@property (readwrite, copy) NSString	*plugType;
@property (readwrite, copy) NSString	*plugCopyright;
@property (readwrite, strong) NSURL		*plugURL;
@end

@implementation PPPlugInInfo
@synthesize authorName;
@synthesize plugName;
@synthesize plugType;
@synthesize plugURL;
@synthesize plugCopyright;

- (id)initWithPlugName:(NSString*)pn author:(NSString*)aut
{
	return [self initWithPlugName:pn author:aut plugType:@"Unknown" plugURL:[[NSBundle mainBundle] bundleURL]];
}

- (id)initWithPlugName:(NSString*)pn author:(NSString*)aut plugType:(NSString*)pt
{
	return [self initWithPlugName:pn author:aut plugType:pt plugURL:[[NSBundle mainBundle] bundleURL]];
}

- (id)initWithPlugName:(NSString*)pn author:(NSString*)aut plugType:(NSString*)pt plugURL:(NSURL*)pu;
{
	if (self = [super init]) {
		NSString *tmpCopyRight = [[NSBundle bundleWithURL:pu] infoDictionary][@"NSHumanReadableCopyright"];
		self.plugName = pn;
		self.authorName = aut;
		self.plugType = pt;
		self.plugURL = pu;
		if (tmpCopyRight) {
			self.plugCopyright = tmpCopyRight;
		} else {
			self.plugCopyright = @"No Copyright Info available";
		}
	}
	return self;
}

- (NSString*)description
{
	return [NSString stringWithFormat:@"%@: %@ (%@), %@", plugName, authorName, plugCopyright, plugType];
}

- (BOOL)isEqual:(id)object
{
	if (!object) {
		return NO;
	}
	if (object == self) {
		return YES;
	}
	if ([object isKindOfClass:[PPPlugInInfo class]]) {
		if (![plugName isEqualToString:[object plugName] ]) {
			return NO;
		} else if (![authorName isEqualToString:[object authorName]]) {
			return NO;
		} else if (![plugType isEqualToString:[object plugType]]) {
			return NO;
		} else
			return YES;
	} else
		return NO;
}

- (NSUInteger)hash
{
	return [plugName hash] ^ [plugType hash] ^ [authorName hash] ^ [plugCopyright hash];
}

@end

File: Shared/PPPlugInInfo.swift
//
//  PPPlugInInfo.swift
//  PPMacho
//
//  Created by C.W. Betts on 7/24/14.
//
//

import Cocoa

func ==(lhs: PPPlugInInfo, rhs: PPPlugInInfo) -> Bool {
	if (lhs === rhs) {
		return true;
	}
	
	if (lhs.plugName != rhs.plugName) {
		return false;
	} else if (lhs.authorName != rhs.authorName) {
		return false;
	} else if (lhs.plugType != rhs.plugType) {
		return false;
	} else {
		return true;
	}
}

class PPPlugInInfo: NSObject, Hashable, DebugPrintable {
	private(set) var plugName: String;
	private(set) var authorName: String;
	private(set) var plugType: String
	private(set) var plugCopyright: String
	private(set) var plugURL: NSURL
	
	init(plugName pn: String, author aut: String, plugType pt: String, plugURL pu: NSURL) {
		var tmpBundle = NSBundle(URL: pu).infoDictionary
		plugName = pn
		authorName = aut
		plugType = pt
		plugURL = pu
		var tmpCopy: AnyObject? = tmpBundle["NSHumanReadableCopyright"]
		if (tmpCopy != nil) {
			plugCopyright = tmpCopy! as String
		} else {
			plugCopyright = "No Copyright Info available"
		}
		super.init()
	}
	
	override var hash: Int { get {
		return self.hashValue
	}}
	
	override var hashValue: Int { get {
		return plugName.hash ^ plugType.hash ^ authorName.hash ^ plugCopyright.hash
	}}
	
	override var debugDescription: String { get {
		return "\(plugName): \(authorName) (\(plugCopyright)), \(plugType)"

	}}
	
	override var description: String { get {
		return self.debugDescription
	}}
	
	override func isEqual(object: AnyObject!) -> Bool {
		if (!object) {
			return false
		}
		if (object === self) {
			return true;
		}
		if (object.isKindOfClass(PPPlugInInfo)) {
			if (plugName != (object as PPPlugInInfo).plugName) {
				return false;
			} else if (authorName != (object as PPPlugInInfo).authorName) {
				return false;
			} else if (plugType != (object as PPPlugInInfo).plugType) {
				return false;
			} else {
				return true;
			}
		} else {
			return false;
		}
	}
}
